### PR TITLE
Add `ToDateTime` parameter attribute

### PR DESCRIPTION
**Pull request by vjik**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌

### CODE REVIEW

File: docs/guide/en/typecasting.md [hunk 1]
```diff
@@ -138,3 +138,20 @@ $money = $hydrator->create(Money::class, [
     'currency' => 'AMD',
 ]);
 ```
+
+To cast a value to `DateTimeImmutable` object explicitly, you can use `ToDateTimeImmutable` attribute:
+
+```php
+use DateTimeImmutable;
+use \Yiisoft\Hydrator\Attribute\Parameter\ToDateTimeImmutable;
```
Comment (xepozz) [inline docs/guide/en/typecasting.md:146]:
```suggestion
use Yiisoft\Hydrator\Attribute\Parameter\ToDateTimeImmutable;
```

File: docs/guide/ru/typecasting.md [hunk 2]
```diff
@@ -138,3 +138,20 @@ $money = $hydrator->create(Money::class, [
     'currency' => 'AMD',
 ]);
 ```
+
+Для приведения значения к объекту `DateTimeImmutable` явно, вы можете использовать атрибут `ToDateTimeImmutable`:
+
+```php
+use DateTimeImmutable;
+use \Yiisoft\Hydrator\Attribute\Parameter\ToDateTimeImmutable;
```
Comment (xepozz) [inline docs/guide/ru/typecasting.md:146]:
```suggestion
use Yiisoft\Hydrator\Attribute\Parameter\ToDateTimeImmutable;
```

File: tests/Attribute/Parameter/ToDateTimeImmutableTest.php [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+            'locale-ru' => [
+                new DateTimeImmutable('12.11.2020, 12:20'),
+                new ToDateTimeImmutable(locale: 'ru'),
+                '12.11.2020, 12:20',
+            ],
+        ];
+    }
+
+    #[DataProvider('dataBase')]
+    public function testBase(mixed $expected, ToDateTimeImmutable $attribute, mixed $value): void
```
Comment (xepozz) [inline tests/Attribute/Parameter/ToDateTimeImmutableTest.php:71]:
```suggestion
    public function testBase(DateTimeImmutable $expected, ToDateTimeImmutable $attribute, mixed $value): void
```

File: tests/Attribute/Parameter/ToDateTimeImmutableTest.php [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+use Yiisoft\Hydrator\Tests\Support\Attribute\CounterResolver;
+use Yiisoft\Hydrator\Tests\Support\Classes\CounterClass;
+use Yiisoft\Hydrator\Tests\Support\TestHelper;
+use Yiisoft\Test\Support\Container\SimpleContainer;
+
+final class ToDateTimeImmutableTest extends TestCase
+{
+    public static function dataBase(): array
+    {
+        return [
```
Comment (xepozz) [inline tests/Attribute/Parameter/ToDateTimeImmutableTest.php:31]: Please use `yield` construction to emit test cases

**Review by xepozz — CHANGES_REQUESTED** [review]
As I can see there is not a lot of modification to make the new attribute be able to convert a value to `DateTime` or `DateTimeImmutable` objects depends on the property type.

Could you make the new attribute to be able to do that?
I'd rename it `ToDateTime` or create a separate attribute.

File: tests/Attribute/Parameter/ToDateTimeImmutableTest.php (hunk 4 above)
Comment (vjik) [reply tests/Attribute/Parameter/ToDateTimeImmutableTest.php]: done

**Comment by vjik** [comment]
> Could you make the new attribute to be able to do that? I'd rename it `ToDateTime` or create a separate attribute.

PHP recommends use `DateTimeImmutable` by default. Is there a case for use mutable `DateTime`?

**Comment by xepozz** [comment]
> > Could you make the new attribute to be able to do that? I'd rename it `ToDateTime` or create a separate attribute.
> 
> PHP recommends use `DateTimeImmutable` by default. Is there a case for use mutable `DateTime`?

Despite on recommendations people still use mutable objects

**Comment by samdark** [comment]
I'd go with immutable and add mutable if/when requested.

**Comment by vjik** [comment]
@xepozz @samdark I added `DateTime` support, but used immutable object, if it avaliable.

See tests: https://github.com/yiisoft/hydrator/pull/77/files#diff-1d0f0e934fc4f20ba9d5353e9be78c218f5aedd536e243089f516d46221d6c44R242

**Review by samdark — APPROVED** [review]

**Pull request merged**